pkg/billing/client.go

```diff
@@ -35,16 +35,150 @@ func GetHostname() string {
 	return hostname
 }
 
-func NewClient(url string, c *http.Client) Client {
-	return Client{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+type Client interface {
+	LogFields() zap.Field
+	send(ctx context.Context, payload []byte, traceID TraceID) error
 }
 
 type TraceID string
 
-func (c Client) GenerateTraceID() TraceID {
+func GenerateTraceID() TraceID {
 	return TraceID(shortuuid.New())
 }
 
+type HTTPClient struct {
+	URL   string
+	httpc *http.Client
+}
+
+func NewHTTPClient(url string, c *http.Client) HTTPClient {
+	return HTTPClient{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+}
+
+func (c HTTPClient) send(ctx context.Context, payload []byte, traceID TraceID) error {
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, c.URL, bytes.NewReader(payload))
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	r.Header.Set("content-type", "application/json")
+	r.Header.Set("x-trace-id", string(traceID))
+
+	resp, err := c.httpc.Do(r)
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	// theoretically if wanted/needed, we should use an http handler that
+	// does the retrying, to avoid writing that logic here.
+	if resp.StatusCode != http.StatusOK {
+		return UnexpectedStatusCodeError{StatusCode: resp.StatusCode}
+	}
+
+	return nil
+}
+
+func (c HTTPClient) LogFields() zap.Field {
+	return zap.String("url", c.URL)
+}
+
+type S3ClientConfig struct {
+	Bucket         string `json:"bucket"`
+	Region         string `json:"region"`
+	PrefixInBucket string `json:"prefixInBucket"`
+	Endpoint       string `json:"endpoint"`
+}
+
+type S3Client struct {
+	cfg    S3ClientConfig
+	client *s3.Client
+}
+
+type S3Error struct {
+	Err error
+}
+
+func (e S3Error) Error() string {
+	return fmt.Sprintf("Error making S3 request: %s", e.Err.Error())
+}
+
+func (e S3Error) Unwrap() error {
+	return e.Err
+}
+
+func NewS3Client(ctx context.Context, cfg S3ClientConfig) (S3Client, error) {
+	s3Config, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(cfg.Region))
+
+	if err != nil {
+		return S3Client{}, S3Error{Err: err} //nolint:exhaustruct // error is returned
```

Comment:
1. the error message for `S3Error` mentions making the request, presumably this shouldn't be wrapped?
2. May be worthwhile to just return `*S3Client` so you can return `nil`. Up to you

Reply:
Fixed in a6a3d3b